### PR TITLE
Correct formatting of preview widget fields

### LINKED ISSUE #3701: UI: Notes section doesn't render line breaks since 2.5.0
## Expected Behavior
Viewable line breaks of notes in the preview section for entries.

## Current Behavior
Linebreaks for notes are only shown correctly when you open an entry to edit it. In the preview section linebreaks are ignored since updating to 2.5.0. Before this update, IIRC, those were rendered as expected.

## Context
Makes it unfeasible to work with notes from the preview section.

## Debug Info
KeePassXC - Version 2.5.0
Revision: 1ab8a9f

Libraries:
- libgcrypt 1.8.5

Operating system: Arch Linux
CPU architecture: x86_64
Kernel: 5.3.7-arch1-1-ARCH

Enabled extensions:
- Auto-Type
- Browser-Integration
- SSH-Agent
- KeeShare (bestätigtes und unbestätigtes Teilen)
- YubiKey

### PR DESCRIPTION
**Pull request by droidmonkey**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change which fixes an issue)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )
* Fix #3701 - replace QLabel with QTextEdit to enable scrolling of notes

* Notes are plain text. They will remain as plain text and hyperlinks will not be enabled in the notes. Until the notes editor is moved to a rich text / html editor this will remain the case.

* Convert username and password fields in preview pane to QLineEdit's to allow for full copying and viewing if larger than the field width.

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )
image

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Tested on Windows

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation, and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

### CODE REVIEW

**Comment by droidmonkey** [comment]
ping @0pLuS0 and @ei-ke and @MyXelf

**Comment by 0pLuS0** [comment]
@droidmonkey HI

If this is for the ```Notes```, hmm since I used 2.4.3, I've personally gotten use to the idea of it having it's own Tab, not sure why it was thought it was a bad idea having it's own tab and now removing it...

Hmm

I think it's best to have the Notes Tab put back in...

**Comment by droidmonkey** [comment]
This is how it is now, we aren't going back. Attributes and Attachments were combined into the Advanced tab.

**Comment by 0pLuS0** [comment]
@droidmonkey ok thanks, I appreciate your time too by the way, keep up the great work!

**Comment by ei-ke** [comment]
@droidmonkey Thank you for the quick fix. Just build your pull request on linux without issues. For the Borders @MyXelf metioned I could live with how it is now. But maybe the background color of the notes field could be kept the save as the rest of the UI?
I've tried this with QT Designer and somehow got it working, but honestly I believe it's better if you do it on your own :)

**Comment by MyXelf** [comment]
As visual aids, here are the screenshots of the 3 variants in Linux (all showing the scrollbars).

Widget with Default Background and StyledPanel (this PR as is):
keepassxc---#3727---preview-notes---01-bgblack-wborders

Widget with Transparent Background and Border:
keepassxc---#3727---preview-notes---02-nobg-wborders

Widget with Transparent Background and No Border (my terrible idea) ;)
keepassxc---#3727---preview-notes---03-nobg-noborders

At the end the visual appearance is OS-specific and user-specific. The light vs dark theme is another issue (that shouldn't interfere, but now it does). And the default look of the `QPlainTextEdit` is definitely suggesting "input", which is not the case.

If requested I could build the screenshots with a light theme.

**Comment by ei-ke** [comment]
@MyXelf now I'd also like the third option: transparent background, no border :)

**Comment by droidmonkey** [comment]
You done convinced me too :sweat_smile:

**Comment by MyXelf** [comment]
:rofl:

The other option to avoid the floating scrollbar is to make the widget adjust to the content of the Notes field and let the "container" do the scrolling if needed. This would be a better way. But right now the Entry Preview is needing a lot of consistency fixes.

I'm about to provide the light theme screenshots.

**Comment by MyXelf** [comment]
Here are the screenshots of the 3 variants in Linux using light theme Breeze (all showing the scrollbars).

Widget with Default Background and StyledPanel (this PR as is):
keepassxc---#3727---preview-notes---light---01-bgwhite-wborders

Widget with Transparent Background and Border:
keepassxc---#3727---preview-notes---light---02-nobg-wborders

Widget with Transparent Background and No Border:
keepassxc---#3727---preview-notes---light---03-nobg-noborders

HTH

**Comment by droidmonkey** [comment]
Hmmm I actually like the idea of scrolling the whole preview widget instead. It makes more sense to me when the border is removed

**Comment by MyXelf** [comment]
Is the `QWidget` able to scroll the content? Or we would need an additional container?

Unless you are able to make the changes in no time it is not worth the effort (IMHO). What's another drop in the ocean? I would focus on the functionality and leave the UX for later (I want to get there eventually). As I said there is a lot that must be done in the Preview UI (misaligned controls, toggle buttons, lot of spacers, inconsistencies between the tabs, to name a few).

**Comment by droidmonkey** [comment]
You would replace the qwidget with a qscrollwidget

**Comment by 0pLuS0** [comment]
@droidmonkey are the Notes always going to visible or will there be some way to hide them when not needed?

**Comment by droidmonkey** [comment]
There is a setting to hide notes by default

**Comment by 0pLuS0** [comment]
@droidmonkey ok thanks, keep up the good work! :+1:

File: src/gui/EntryPreviewWidget.cpp [hunk 1]
```diff
@@ -189,28 +189,31 @@ void EntryPreviewWidget::setPasswordVisible(bool state)
...
 {
+    notesWidget->clear();
+
     if (state) {
         // Add html hyperlinks to notes that start with XXXX://
         QString hyperlinkNotes = notes;
-        notesLabel->setText(hyperlinkNotes.replace(QRegExp("(\\w+:\\/\\/\\S+)"), "<a href=\"\\1\">\\1</a>"));
-        notesLabel->setTextInteractionFlags(Qt::TextBrowserInteraction);
+        hyperlinkNotes.replace("\n", "<br/>");
+        notesWidget->appendHtml(hyperlinkNotes.replace(QRegExp("(\\w+:\\/\\/\\S+)"), "<a href=\"\\1\">\\1</a>"));
```
Comment (phoerious) [inline src/gui/EntryPreviewWidget.cpp]: hyperlinkNotes needs proper HTML escaping before doing any replacements and then appending it as HTML to the widget.

Comment (droidmonkey) [reply src/gui/EntryPreviewWidget.cpp]: Good point

Comment (droidmonkey) [reply src/gui/EntryPreviewWidget.cpp]: I also need to fix the styling of this

**Comment by 0pLuS0** [comment]
Well the DEV(s) are at this PLEASE make the ```Password Preview``` show the entire password, which it doesn't on passwords with long characters...

Keep up the great work!

THANKS

**Comment by droidmonkey** [comment]
Will do

**Comment by 0pLuS0** [comment]
> Will do

:+1:

**Comment by droidmonkey** [comment]
This is now final and addresses all concerns. One caveat is that the hyperlinks in the notes have been disabled. I spent a good 3 hours trying to satisfy all criteria for our notes (allow monospace font, obey whitespace, etc. etc.) and adding hyperlinks not only adds a ton of code, but it also introduces several edge cases due to parsing of the notes.

To this end, when notes become a rich text / html editor we can bring hyperlinks back into the fold.

**Pull request merged**